In [1]:
import pandas as pd

In [3]:
df = pd.read_csv("university.csv")
df

,University Rank,Name of University,Location,No of student,No of student per staff,International Student,Female:Male Ratio,OverAll Score,Teaching Score,Research Score,Citations Score,Industry Income Score,International Outlook Score
0,1,University of Oxford,United Kingdom,"20,965",10.6,42%,48 : 52,96.4,92.3,99.7,99.0,74.9,96.2
1,2,Harvard University,United States,"21,887",9.6,25%,50 : 50,95.2,94.8,99.0,99.3,49.5,80.5
2,3,University of Cambridge,United Kingdom,"20,185",11.3,39%,47 : 53,94.8,90.9,99.5,97.0,54.2,95.8
3,3,Stanford University,United States,"16,164",7.1,24%,46 : 54,94.8,94.2,96.7,99.8,65.0,79.8
4,5,Massachusetts Institute of Technology,United States,"11,415",8.2,33%,40 : 60,94.2,90.7,93.6,99.8,90.9,89.3
...,...,...,...,...,...,...,...,...,...,...,...,...,...
2336,-,University of the West of Scotland,NaN,NaN,NaN,NaN,NaN,34.0–39.2,24.1,15.5,61.5,37.9,76.8
2337,-,University of Windsor,NaN,NaN,NaN,NaN,NaN,34.0–39.2,35.1,29.4,34.5,44.2,88.7
2338,-,University of Wolverhampton,NaN,NaN,NaN,NaN,NaN,34.0–39.2,18.2,14.3,68.8,37.3,72.0
2339,-,University of Wuppertal,NaN,NaN,NaN,NaN,NaN,34.0–39.2,26.4,26.7,52.8,52.1,47.6


In [4]:
df.count()

University Rank                2341
Name of University             2233
Location                       2047
No of student                  2209
No of student per staff        2208
International Student          2209
Female:Male Ratio              2128
OverAll Score                  1799
Teaching Score                 1799
Research Score                 1799
Citations Score                1799
Industry Income Score          1799
International Outlook Score    1799
dtype: int64

### 📝 Answers to the Assignment Questions

1. **How many universities are there in the dataset?**
   There are **2,233** unique universities in the dataset.

2. **How many different countries are there?**
   There are **116** different countries represented.

3. **What is the distribution of countries in the top 100 universities?**
   *(Note: 26 universities in the top 100 are missing their Location value in this dataset)*
   - United States: 34
   - United Kingdom: 7
   - Germany: 6
   - Australia: 4
   - Switzerland: 4
   - Netherlands: 4
   - France: 4
   - China: 3
   - Canada: 2
   - Japan: 2
   - South Korea: 2
   - Sweden: 1
   - Singapore: 1

4. **What is the average number of students in the top 10 universities?**
   The average number of students in the top 10 universities is **17,408**.

5. **Which are the top 10 universities with more than 50% international student?**
   1. Imperial College London (61%)
   2. UCL (60%)
   3. King’s College London (52%)
   4. London School of Economics and Political Science (73%)
   5. École Polytechnique Fédérale de Lausanne (62%)
   6. Australian National University (51%)
   7. City University of Hong Kong (62%)
   8. Maastricht University (56%)
   9. University of Luxembourg (51%)
   10. Macau University of Science and Technology (91%)

6. **Which are the top 10 universities with a predominantly female presence?**
   *(Defined as having a female ratio > 50%)*
   1. University of California, Berkeley (52%)
   2. Yale University (52%)
   3. University of Pennsylvania (53%)
   4. Johns Hopkins University (53%)
   5. University of Toronto (56%)
   6. National University of Singapore (51%)
   7. Cornell University (51%)
   8. University of California, Los Angeles (56%)
   9. UCL (59%)
   10. New York University (57%)

In [6]:
import numpy as np

# Clean 'No of student' (remove commas and convert to float)
df['No of student'] = df['No of student'].astype(str).str.replace(',', '').astype(float)

# Clean 'International Student' (remove % and convert to numeric)
df['International Student'] = df['International Student'].astype(str).str.replace('%', '')
df['International Student'] = pd.to_numeric(df['International Student'], errors='coerce')

# Clean 'Female:Male Ratio' (extract the female percentage)
def get_female_ratio(ratio_str):
    if pd.isna(ratio_str):
        return np.nan
    try:
        female_part = str(ratio_str).split(':')[0].strip()
        return float(female_part)
    except:
        return np.nan

df['Female Ratio'] = df['Female:Male Ratio'].apply(get_female_ratio)

In [7]:
# 1. How many universities are there in the dataset?
num_universities = df['Name of University'].nunique()
print(f"1. Number of unique universities: {num_universities}")

# 2. How many different countries are there?
num_countries = df['Location'].nunique()
print(f"2. Number of different countries: {num_countries}")

# 3. What is the distribution of countries in the top 100 universities?
top_100 = df.head(100)
dist_top_100 = top_100['Location'].value_counts()
print(f"\n3. Distribution of countries in top 100 universities:\n{dist_top_100.to_string()}")

# 4. What is the average number of students in the top 10 universities?
avg_students_top_10 = df.head(10)['No of student'].mean()
print(f"\n4. Average number of students in top 10 universities: {avg_students_top_10:,.0f}")

# 5. Which are the top 10 universities with more than 50% international student?
over_50_int = df[df['International Student'] > 50]
top_10_over_50_int = over_50_int.head(10)[['University Rank', 'Name of University', 'International Student']]
print(f"\n5. Top 10 universities with >50% international students:\n{top_10_over_50_int.to_string(index=False)}")

# 6. Which are the top 10 universities with a predominantly female presence?
predominantly_female = df[df['Female Ratio'] > 50]
top_10_female = predominantly_female.head(10)[['University Rank', 'Name of University', 'Female:Male Ratio']]
print(f"\n6. Top 10 universities with predominantly female presence:\n{top_10_female.to_string(index=False)}")

1. Number of unique universities: 2233
2. Number of different countries: 116

3. Distribution of countries in top 100 universities:
Location
United States     34
United Kingdom     7
Germany            6
Australia          4
Switzerland        4
Netherlands        4
France             4
China              3
Canada             2
Japan              2
South Korea        2
Sweden             1
Singapore          1

4. Average number of students in top 10 universities: 17,408

5. Top 10 universities with >50% international students:
University Rank                               Name of University  International Student
             10                          Imperial College London                   61.0
             22                                              UCL                   60.0
             35                            King’s College London                   52.0
             37 London School of Economics and Political Science                   73.0
             41         Éc